pkg/storage/chunk/client/hedging/hedging_test.go

```diff
@@ -18,37 +18,39 @@ func (fn RoundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error)
 	return fn(req)
 }
 
-func resetMetrics() {
+func resetMetrics() *prometheus.Registry {
+	//TODO: clean up this massive hack...
 	reg := prometheus.NewRegistry()
 	prometheus.DefaultRegisterer = reg
 	prometheus.DefaultGatherer = reg
 	initMetrics()
+	return reg
 }
 
 func TestHedging(t *testing.T) {
-	resetMetrics()
+	reg := resetMetrics()
```

Comment:
This code is quite gnarly and it's not easy to clean up without basically rewriting it... :disappointed: 

The problem is that the tests rely on the actual values of the prometheus metrics (see the `require.NoError(t, testutil.GatherAndCompare` parts of both tests). And while the actual code relies on an injected registry, it also [relies on a global `sync.Once`](https://github.com/grafana/loki/blob/ad85cc2276a42d899fe27b27f5940550f47bcc21/pkg/storage/chunk/client/hedging/hedging.go#L105-L109) that it's not clear I can remove, because `ClientWithRegisterer()` and `RoundTripperWithRegisterer()` are called in a bunch of different places (see [this](https://github.com/search?q=repo%3Agrafana%2Floki%20ClientWithRegisterer&type=code) and [this](https://github.com/search?q=repo%3Agrafana%2Floki+RoundTripperWithRegisterer&type=code)) :confused: 

So yeah, to not turn this into a whole project, I intentionally chose to just fix the tests to run fine with `-race` and/or `-count 2`, with the minimal amount of changes to resolve the immediate issue, without touching the actual code. Though I do think the rest of the code probably needs to be rewritten, just not now...

Reply:
ah thanks for the further info, I was hoping there wouldn't be hidden calls to the global registry in there.... thanks for looking and explaining!